src/main/java/software/amazon/smithy/lsp/Utils.java

```diff
@@ -37,4 +44,42 @@ public U get() {
 
     return CompletableFuture.supplyAsync(supplier);
   }
+
+  public static Boolean isSmithyJarFile(String rawUri) throws IOException {
+    try {
+      String uri = java.net.URLDecoder.decode(rawUri, StandardCharsets.UTF_8.name());
+      return uri.startsWith("smithyjar:");
+    } catch (IOException e) {
+      return false;
+    }
+  }
+
+  public static Boolean isFile(String rawUri) {
+    try {
+      String uri = java.net.URLDecoder.decode(rawUri, StandardCharsets.UTF_8.name());
+      return uri.startsWith("file:");
+    } catch (IOException e) {
+      return false;
+    }
+  }
+
+  public static List<String> jarFileContents(String rawUri, ClassLoader classLoader) throws IOException {
+    if (Utils.isSmithyJarFile(rawUri)) {
+      String uri = java.net.URLDecoder.decode(rawUri, StandardCharsets.UTF_8.name());
+      String[] pathArray = uri.split("!/");
+      String resourcePath = pathArray[1];
+      InputStream is = classLoader.getResourceAsStream(resourcePath);
+      try {
+        if (is != null) {
+          InputStreamReader isr = new InputStreamReader(is);
+          BufferedReader reader = new BufferedReader(isr);
+          return reader.lines().collect(Collectors.toList());
+        }
+      } finally {
+        is.close();
+      }
+    }
+    return null;
```

Comment:
Returning an empty list here would prevent null pointer exceptions later down the line. Alternatively (and probably the better idea), throw an exception since the input was invalid.